### PR TITLE
MSC3873: Escape keys when flattening dicts.

### PR DESCRIPTION
**Pull request by clokep**
This is a potential implementation for [MSC3873](https://github.com/matrix-org/matrix-spec-proposals/pull/3873), ~~it doesn't quite match the MSC right now due to implementing a simplification of it.~~ (The MSC was updated to match this.)

~~Based on #15002.~~

### CODE REVIEW

File: rust/src/push/base_rules.rs [hunk 1]
```diff
@@ -71,7 +71,7 @@ pub const BASE_APPEND_OVERRIDE_RULES: &[PushRule] = &[
         priority_class: 5,
         conditions: Cow::Borrowed(&[Condition::Known(KnownCondition::EventMatch(
             EventMatchCondition {
-                key: Cow::Borrowed("content.m.relates_to.rel_type"),
+                key: Cow::Borrowed("content.m\\.relates_to.rel_type"),
```
Comment (clokep) [inline rust/src/push/base_rules.rs:74]: I think this is the only reasonable thing to do here -- make it abide by the new behavior?

Comment (clokep) [reply rust/src/push/base_rules.rs:74]: It does make it hard to have this behind a configuration variable though? I'm not really sure how to handle this.

Thoughts are welcome!

Comment (squahtx) [reply rust/src/push/base_rules.rs:74]: Is this change visible to users? I suppose we could represent keys unambiguously inside Synapse and degrade them to ambiguous form when matching / presenting them to users and the experimental flag is disabled.

Then that raises the question of how custom push rules would work when a server admin toggles the experimental flag. We could also record whether a key is in msc3873 format, but that would add a bunch of complexity to a fairly simple PR.

**Review by squahtx — APPROVED** [review]
The current changes look good. Gating it behind an experimental flag would be nice but sounds like a lot of faff. If you can see a nice way to go about it, I'd go for it.

File: rust/src/push/base_rules.rs (hunk 1 above)
Comment (clokep) [reply rust/src/push/base_rules.rs:74]: > Is this change visible to users?

No, this rule is disabled by default. Mostly Synapse needs to be internally consistent here?

> I suppose we could represent keys unambiguously inside Synapse and degrade them to ambiguous form when matching / presenting them to users and the experimental flag is disabled.

The MSC originally had this and we removed it.

**Comment by clokep** [comment]
> Gating it behind an experimental flag would be nice but sounds like a lot of faff. If you can see a nice way to go about it, I'd go for it.

I did this, unfortunately if you enable both it + MSC3958 then that one won't work properly. I think this is "OK" for experimental stuff though?

**Comment by squahtx** [comment]
\> I did this, unfortunately if you enable both it + [MSC3958](https://github.com/matrix-org/matrix-spec-proposals/pull/3958) then that one won't work properly. I think this is "OK" for experimental stuff though?

Sounds okay to have those experimental features be mutually exclusive? I'm mildly concerned that we'll forget to make them work together when stabilizing them.

**Comment by clokep** [comment]
> Sounds okay to have those experimental features be mutually exclusive? I'm mildly concerned that we'll forget to make them work together when stabilizing them.

There's a test that will fail if we forget! 😄

File: synapse/config/experimental.py [hunk 2]
```diff
@@ -169,6 +169,11 @@ def read_config(self, config: JsonDict, **kwargs: Any) -> None:
         # MSC3925: do not replace events with their edits
         self.msc3925_inhibit_edit = experimental.get("msc3925_inhibit_edit", False)
 
+        # MSC3873: Disambiguate event_match keys.
+        self.msc3783_escape_event_match_key = experimental.get(
+            "msc3783_escape_event_match_key", False
```
Comment (SuperSandro2000) [inline synapse/config/experimental.py:174]: Shouldn't this be this?

```suggestion
        self.msc3783_escape_event_match_key = experimental.get(
            "msc3783_escape_event_match_key", False
```

Comment (clokep) [reply synapse/config/experimental.py:174]: Sorry -- I don't see any change in your suggestion?

Comment (SuperSandro2000) [reply synapse/config/experimental.py:174]: 😅 did the same typo. Now corrected.

Comment (clokep) [reply synapse/config/experimental.py:174]: Oh the MSC # is typoed, I see. Umm, yeah that's not awesome, although not super important. 😢

Comment (clokep) [reply synapse/config/experimental.py:174]: See #15138

**Pull request merged**